### PR TITLE
Standardize "operation successful" messages

### PR DESCRIPTION
**Pull request by dregad**
Adding new API functions to handle display of confirmation messages, and replace custom HTML in various files to display the same by calls to these functions.

Fixes [#21683](https://www.mantisbt.org/bugs/view.php?id=21683)

### CODE REVIEW

File: core/html_api.php [hunk 1]
```diff
@@ -369,27 +369,98 @@ function html_top_banner() {
...
- * A function that outputs that an operation was successful and provides a redirect link.
- * @param string $p_redirect_url The url to redirect to.
- * @param string $p_message      Message to display to the user.
+ * Outputs an operation successful message with multiple redirection buttons.
+ * @param array  $p_buttons      Array of (URL, label) pairs used to generate
+ *                               the buttons; if label is null or unspecified,
+ *                               the default 'proceed' text will be displayed.
+ * @param string $p_message      Message to display to the user. If none is
+ *                               provided, a default message will be printed
+ * @param string $p_type         One of the constants CONFIRMATION_TYPE_SUCCESS,
```
Comment (atrol) [inline core/html_api.php]: integer $p_type

File: core/html_api.php [hunk 2]
```diff
@@ -369,27 +369,98 @@ function html_top_banner() {
...
  */
-function html_operation_successful( $p_redirect_url, $p_message = '' ) {
+function html_operation_successful_buttons( array $p_buttons, $p_message = '', $p_type = CONFIRMATION_TYPE_SUCCESS ) {
+	switch( $p_type ) {
+		case CONFIRMATION_TYPE_FAILURE:
+			$t_alert_css = 'alert-danger';
+			$t_message = 'operation_failed';
+			break;
+		case CONFIRMATION_TYPE_WARNING:
+			$t_message = 'operation_warnings';
```
Comment (atrol) [inline core/html_api.php]: For better readability and to be consistent with other cases.
Insert this line after `$t_alert_css = 'alert-warning';`

File: core/html_api.php [hunk 3]
```diff
@@ -369,27 +369,98 @@ function html_top_banner() {
...
+ *                               the buttons; if label is null or unspecified,
+ *                               the default 'proceed' text will be displayed.
+ * @param string $p_message      Message to display to the user. If none is
+ *                               provided, a default message will be printed
+ * @param string $p_type         One of the constants CONFIRMATION_TYPE_SUCCESS,
+ *                               CONFIRMATION_TYPE_WARNING, CONFIRMATION_TYPE_FAILURE
  * @return void
  */
-function html_operation_successful( $p_redirect_url, $p_message = '' ) {
+function html_operation_successful_buttons( array $p_buttons, $p_message = '', $p_type = CONFIRMATION_TYPE_SUCCESS ) {
```
Comment (atrol) [inline core/html_api.php]: html_operation_confirmation_buttons might be a better name for the function.
It's a bit strange to to have the word successful for failures and warnings, e.g.
```php
function html_operation_failure( $p_redirect_url, $p_message = '' ) {
	html_operation_successful_buttons(...
```

File: manage_user_create.php [hunk 4]
```diff
@@ -131,16 +131,12 @@
 layout_page_header( null, $t_redirect_url );
 
 layout_page_begin( 'manage_overview_page.php' );
-?>
-
-<div class="space-20"></div>
-<div class="alert alert-success">
-<?php
-$t_access_level = get_enum_element( 'access_levels', $f_access_level );
```
Comment (atrol) [inline manage_user_create.php]: Don't remove this line as you use $t_access_level to build $t_message

File: manage_user_reset.php [hunk 5]
```diff
@@ -79,31 +79,24 @@
...
-echo '<div class="col-md-12 col-xs-12">';
-echo '<div class="space-10"></div>';
-echo '<div class="alert alert-success">';
-
-echo '<p class="bigger-110">';
 if( $t_reset ) {
 	if( false == $t_result ) {
 		# PROTECTED
-		echo lang_get( 'account_reset_protected_msg' );
+		# @TODO this should be a failure message !
```
Comment (atrol) [inline manage_user_reset.php]: So why don't you use html_operation_failure?

File: manage_user_update.php [hunk 6]
```diff
@@ -226,12 +226,12 @@
...
 
-if( $f_protected && $t_old_protected ) {				# PROTECTED
-	echo '<div class="failure-msg">';
-	echo lang_get( 'manage_user_protected_msg' ) . '<br />';
-	print_button( $t_redirect_url, lang_get( 'proceed' ) );
-	echo '</div>';
-} else if( $t_result ) {					# SUCCESS
+if( $f_protected && $t_old_protected ) {
+	# PROTECTED
+	# @TODO this should actually be a warning message
```
Comment (atrol) [inline manage_user_update.php]: So why don't you use html_operation_warning?

**Comment by atrol** [comment]
👍 in general, just reviewed, didn't test.

File: core/html_api.php (hunk 1 above)
Comment (dregad) [inline core/html_api.php]: >integer $p_type

Right, forgot to change the phpdoc when I moved the strings to constants

File: core/html_api.php (hunk 2 above)
Comment (dregad) [inline core/html_api.php]: > Insert this line after `$t_alert_css = 'alert-warning';`

Ok.

File: core/html_api.php (hunk 3 above)
Comment (dregad) [inline core/html_api.php]: > html_operation_confirmation_buttons might be a better name for the function.

Good point. Initially, this function was only to extend the existing `html_operation_successful()` function, then I realized there were use cases for warning/failure but forgot to change the name. Will do.

File: manage_user_create.php (hunk 4 above)
Comment (dregad) [inline manage_user_create.php]: Oops...

File: manage_user_reset.php (hunk 5 above)
Comment (dregad) [inline manage_user_reset.php]: I was planning to (hence the @TODO), then forgot about it...

File: manage_user_update.php (hunk 6 above)
Comment (dregad) [inline manage_user_update.php]: > So why don't you use html_operation_warning?

Same comment as above

**Comment by dregad** [comment]
Thanks @atrol for the review, I made the necessary corrections.

File: bug_report.php [hunk 7]
```diff
@@ -322,16 +322,11 @@
...
-	<div class="alert alert-success">
 
-<?php
-echo '<p>' . lang_get( 'operation_successful' ) . '</p><br />';
-print_button( string_get_bug_view_url( $t_bug_id ), sprintf( lang_get( 'view_submitted_bug_link' ), $t_bug_id ) );
-print_button( 'view_all_bug_page.php', lang_get( 'view_bugs_link' ) );
+html_operation_confirmation( array(
+	array( string_get_bug_view_url( $t_bug_id ), sprintf( lang_get( 'view_submitted_bug_link' ), $t_bug_id ) ),
+	array( 'view_all_bug_page.php', lang_get( 'view_bugs_link' ) ),
+) );
```
Comment (syncguru) [inline bug_report.php]: Should this be a call to html_operation_successful instead?
On a side note, I find this function call with array of arrays input to be rather complex. I would suggest separating the buttons build up logic first. Also using static arrays if possible http://php.net/SplFixedArray

File: news_update.php [hunk 8]
```diff
@@ -69,17 +69,10 @@
...
-print_button( 'news_edit_page.php?news_id=' . $f_news_id . '&action=edit', lang_get( 'edit_link' ) );
-echo '&nbsp;&nbsp;&nbsp;&nbsp;';
-print_button( 'news_menu_page.php', lang_get( 'proceed' ) );
-echo '</div>';
-echo '</div>';
-echo '</div>';
+html_operation_confirmation( array(
+	array( 'news_menu_page.php' ),
+	array( 'news_edit_page.php?news_id=' . $f_news_id . '&action=edit', lang_get( 'edit_link' ) ),
+) );
```
Comment (syncguru) [inline news_update.php]: Same here.

**Comment by syncguru** [comment]
Apart from comments, this PR looks good.

File: bug_report.php (hunk 7 above)
Comment (dregad) [inline bug_report.php]: > Should this be a call to html_operation_successful instead?

html_operation_successful() only allows a single button, so to maintain existing functionality it is required to use html_operation_confirmation().

> I find this function call with array of arrays input to be rather complex. 
> I would suggest separating the buttons build up logic first. 

So you're saying I should do this instead ?
```
$t_buttons = array(
	array( string_get_bug_view_url( $t_bug_id ), sprintf( lang_get( 'view_submitted_bug_link' ), $t_bug_id ) ),
	array( 'view_all_bug_page.php', lang_get( 'view_bugs_link' ) ),
);
html_operation_confirmation( $t_buttons );
```
If that's the case, I don't really see the benefit.

> Also using static arrays if possible http://php.net/SplFixedArray

Considering that in 100% of current use cases the array size is <= 2, and that it's not reused after the function call, I'm not sure there is a real benefit for fixed-size arrays here.

File: news_update.php (hunk 8 above)
Comment (dregad) [inline news_update.php]: Same reply...

File: bug_report.php (hunk 7 above)
Comment (syncguru) [inline bug_report.php]: The code snippet is more clear than the original.

**Comment by syncguru** [comment]
👍

**Comment by dregad** [comment]
@syncguru thanks for the feedback. For the record I did not use SplFixedArray because I don't think it makes sense in this context, it just adds unnecessary overhead.

I just rebased the branch to fix the conflict on _lost_pwd.php_

**Comment by cproensa** [comment]
Looks good, but i could only test lightly so far.

By the way, do you plan to port it to 1.3?

**Comment by dregad** [comment]
> do you plan to port it to 1.3?

I was not planning to initially, but I guess it would make sense

File: lang/strings_english.txt [hunk 9]
```diff
@@ -103,6 +103,8 @@ $s_actiongroup_error_issue_is_readonly = 'Issue is readonly.';
 $s_all_projects = 'All Projects';
 $s_move_bugs = 'Move Issues';
 $s_operation_successful = 'Operation successful.';
+$s_operation_warnings = 'Operation completed with warnings.';
+$s_operation_failed = 'Operation did not complete successfully.';
```
Comment (vboctor) [inline lang/strings_english.txt]: Why not "Operation failed."?  Do we or should we include reasons when it fails?  Not a big deal.

**Comment by jllano** [comment]
@dregad 
1. I have had checkout the master branch, able to see your changes
2. I tried to replicate the issue on #894 , and got this below:

checkstay_broken

**Comment by dregad** [comment]
Opened [#21710](https://www.mantisbt.org/bugs/view.php?id=21710) to track this

**Pull request merged**